src/plugins/intel_npu/src/backend/src/zero_infer_request.cpp

```diff
@@ -703,6 +810,47 @@ std::vector<ov::ProfilingInfo> ZeroInferRequest::get_profiling_info() const {
     }
 }
 
+std::shared_ptr<ov::ITensor> ZeroInferRequest::allocate_tensor(const IODescriptor& descriptor,
+                                                               const size_t index,
+                                                               const bool isInput,
+                                                               const ov::Allocator& allocator,
+                                                               const std::optional<std::size_t> batchSize) const {
+    check_network_precision(descriptor.precision);
+
+    std::shared_ptr<ov::ITensor> tensor;
+    ov::Shape allocatedTensorShape = descriptor.shapeFromCompiler.get_max_shape();
+
+    if (batchSize.has_value()) {
+        allocatedTensorShape[BATCH_AXIS] = *batchSize;
+    }
+
+    if (descriptor.isStateOutput) {
+        // Only one buffer is required for each (state input, state output) pair, acting as an input before running the
+        // inference and as an output after performing it. Thus both the "state input" and "state output" entries shall
+        // point to the same buffer.
+        OPENVINO_ASSERT(descriptor.relatedDescriptorIndex.has_value(),
+                        "The link between state descriptors is missing, state name: ",
+                        descriptor.nameFromCompiler);
+        tensor = get_user_input(*descriptor.relatedDescriptorIndex)._ptr;
+    } else {
+        tensor = std::make_shared<ZeroTensor>(_initStructs, descriptor.precision, allocatedTensorShape, allocator);
```

Comment:
Or maybe you could define a `make_tensor` virtual method inside `SyncInferRequest` and override it for L0 as to return `ZeroTensor` instead of the previous `make_tensor` call. This way you won't have to copy almost the whole code for `allocate_tensor`. Just something to consider.